## Load Data

In [1]:
%pip install -q kagglehub

from pathlib import Path
import pandas as pd
import kagglehub

# Download the latest Olist Brazilian E-Commerce dataset from Kaggle.
dataset_path = Path(
    kagglehub.dataset_download("olistbr/brazilian-ecommerce")
)

# Locate the folder containing the CSV files.
# This keeps RAW_DIR working the same way as in the original notebook.
orders_file = next(dataset_path.rglob("olist_orders_dataset.csv"))
RAW_DIR = orders_file.parent

csv_files = sorted(RAW_DIR.glob("*.csv"))

print(f"Found {len(csv_files)} CSV files:")
for file in csv_files:
    print("-", file.name)


Using Colab cache for faster access to the 'brazilian-ecommerce' dataset.
Found 9 CSV files:
- olist_customers_dataset.csv
- olist_geolocation_dataset.csv
- olist_order_items_dataset.csv
- olist_order_payments_dataset.csv
- olist_order_reviews_dataset.csv
- olist_orders_dataset.csv
- olist_products_dataset.csv
- olist_sellers_dataset.csv
- product_category_name_translation.csv


## General Table Inspection

In [2]:
for file in csv_files:
    df = pd.read_csv(file)

    print("\n" + "=" * 80)
    print(file.name)
    print("=" * 80)

    print(f"Rows: {df.shape[0]:,}")
    print(f"Columns: {df.shape[1]}")

    print("\nColumn names:")
    print(df.columns.tolist())

    print("\nData types:")
    print(df.dtypes)

    print("\nMissing values:")
    missing = pd.DataFrame({
        "missing_count": df.isna().sum(),
        "missing_percent": (df.isna().mean() * 100).round(2)
    })
    print(missing[missing["missing_count"] > 0])

    print(f"\nExact duplicate rows: {df.duplicated().sum():,}")

    print("\nFirst 3 rows:")
    display(df.head(3))

    del df


olist_customers_dataset.csv
Rows: 99,441
Columns: 5

Column names:
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

Data types:
customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

Missing values:
Empty DataFrame
Columns: [missing_count, missing_percent]
Index: []

Exact duplicate rows: 0

First 3 rows:


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP



olist_geolocation_dataset.csv
Rows: 1,000,163
Columns: 5

Column names:
['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']

Data types:
geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                object
geolocation_state               object
dtype: object

Missing values:
Empty DataFrame
Columns: [missing_count, missing_percent]
Index: []

Exact duplicate rows: 261,831

First 3 rows:


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP



olist_order_items_dataset.csv
Rows: 112,650
Columns: 7

Column names:
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']

Data types:
order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

Missing values:
Empty DataFrame
Columns: [missing_count, missing_percent]
Index: []

Exact duplicate rows: 0

First 3 rows:


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.0,17.87



olist_order_payments_dataset.csv
Rows: 103,886
Columns: 5

Column names:
['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']

Data types:
order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

Missing values:
Empty DataFrame
Columns: [missing_count, missing_percent]
Index: []

Exact duplicate rows: 0

First 3 rows:


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71



olist_order_reviews_dataset.csv
Rows: 99,224
Columns: 7

Column names:
['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']

Data types:
review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object

Missing values:
                        missing_count  missing_percent
review_comment_title            87656            88.34
review_comment_message          58247            58.70

Exact duplicate rows: 0

First 3 rows:


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24



olist_orders_dataset.csv
Rows: 99,441
Columns: 8

Column names:
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

Data types:
order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

Missing values:
                               missing_count  missing_percent
order_approved_at                        160             0.16
order_delivered_carrier_date            1783             1.79
order_delivered_customer_date           2965             2.98

Exact duplicate rows: 0

First 3 rows:


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00



olist_products_dataset.csv
Rows: 32,951
Columns: 9

Column names:
['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']

Data types:
product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

Missing values:
                            missing_count  missing_percent
product_category_name                 610             1.85
product_name_lenght                   610             1.85
product_description_lenght            610             1.85
product_photos_qty                    610             1.85
product_weight_g                        2             0

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0



olist_sellers_dataset.csv
Rows: 3,095
Columns: 4

Column names:
['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']

Data types:
seller_id                 object
seller_zip_code_prefix     int64
seller_city               object
seller_state              object
dtype: object

Missing values:
Empty DataFrame
Columns: [missing_count, missing_percent]
Index: []

Exact duplicate rows: 0

First 3 rows:


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ



product_category_name_translation.csv
Rows: 71
Columns: 2

Column names:
['product_category_name', 'product_category_name_english']

Data types:
product_category_name            object
product_category_name_english    object
dtype: object

Missing values:
Empty DataFrame
Columns: [missing_count, missing_percent]
Index: []

Exact duplicate rows: 0

First 3 rows:


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto


## Key Checks

In [3]:
key_checks = {
    "olist_customers_dataset.csv": ["customer_id"],
    "olist_orders_dataset.csv": ["order_id"],
    "olist_order_items_dataset.csv": ["order_id", "order_item_id"],
    "olist_products_dataset.csv": ["product_id"],
    "olist_sellers_dataset.csv": ["seller_id"],
    "olist_order_payments_dataset.csv": ["order_id", "payment_sequential"],
    "olist_order_reviews_dataset.csv": ["review_id"],
    "product_category_name_translation.csv": ["product_category_name"],
}

for filename, keys in key_checks.items():
    file = RAW_DIR / filename

    if not file.exists():
        print(f"{filename}: FILE NOT FOUND")
        continue

    df = pd.read_csv(file)

    duplicates = df.duplicated(subset=keys).sum()

    print(
        f"{filename}: "
        f"{keys} -> {duplicates:,} duplicate key combination(s)"
    )

    del df

olist_customers_dataset.csv: ['customer_id'] -> 0 duplicate key combination(s)
olist_orders_dataset.csv: ['order_id'] -> 0 duplicate key combination(s)
olist_order_items_dataset.csv: ['order_id', 'order_item_id'] -> 0 duplicate key combination(s)
olist_products_dataset.csv: ['product_id'] -> 0 duplicate key combination(s)
olist_sellers_dataset.csv: ['seller_id'] -> 0 duplicate key combination(s)
olist_order_payments_dataset.csv: ['order_id', 'payment_sequential'] -> 0 duplicate key combination(s)
olist_order_reviews_dataset.csv: ['review_id'] -> 814 duplicate key combination(s)
product_category_name_translation.csv: ['product_category_name'] -> 0 duplicate key combination(s)


## Date/Time Columns

In [4]:
for file in csv_files:
    df = pd.read_csv(file, nrows=5)

    date_columns = [
        col for col in df.columns
        if "date" in col.lower() or "timestamp" in col.lower()
    ]

    if date_columns:
        print(f"{file.name}:")
        for col in date_columns:
            print("  -", col)

olist_order_items_dataset.csv:
  - shipping_limit_date
olist_order_reviews_dataset.csv:
  - review_creation_date
  - review_answer_timestamp
olist_orders_dataset.csv:
  - order_purchase_timestamp
  - order_delivered_carrier_date
  - order_delivered_customer_date
  - order_estimated_delivery_date


## Duplicate Reviews

In [5]:
reviews = pd.read_csv(RAW_DIR / "olist_order_reviews_dataset.csv")

duplicate_review_ids = reviews[
    reviews.duplicated(subset=["review_id"], keep=False)
].sort_values("review_id")

print("Rows involved in duplicated review IDs:", len(duplicate_review_ids))
print("Unique duplicated review IDs:", duplicate_review_ids["review_id"].nunique())

display(duplicate_review_ids.head(20))

review_id_order_counts = (
    reviews.groupby("review_id")["order_id"]
    .nunique()
    .sort_values(ascending=False)
)

print(review_id_order_counts.head(10))

Rows involved in duplicated review IDs: 1603
Unique duplicated review IDs: 789


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
46678,00130cbe1f9d422698c812ed8ded1919,dfcdfc43867d1c1381bfaf62d6b9c195,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07 00:00:00,2018-03-20 18:08:23
29841,00130cbe1f9d422698c812ed8ded1919,04a28263e085d399c97ae49e0b477efa,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07 00:00:00,2018-03-20 18:08:23
90677,0115633a9c298b6a98bcbe4eee75345f,78a4201f58af3463bdab842eea4bc801,5,NaN,NaN,2017-09-21 00:00:00,2017-09-26 03:27:47
63193,0115633a9c298b6a98bcbe4eee75345f,0c9850b2c179c1ef60d2855e2751d1fa,5,NaN,NaN,2017-09-21 00:00:00,2017-09-26 03:27:47
92876,0174caf0ee5964646040cd94e15ac95e,f93a732712407c02dce5dd5088d0f47b,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07 00:00:00,2018-03-08 03:00:53
57280,0174caf0ee5964646040cd94e15ac95e,74db91e33b4e1fd865356c89a61abf1f,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07 00:00:00,2018-03-08 03:00:53
54832,017808d29fd1f942d97e50184dfb4c13,8daaa9e99d60fbba579cc1c3e3bfae01,5,NaN,NaN,2018-03-02 00:00:00,2018-03-05 01:43:30
99167,017808d29fd1f942d97e50184dfb4c13,b1461c8882153b5fe68307c46a506e39,5,NaN,NaN,2018-03-02 00:00:00,2018-03-05 01:43:30
20621,0254bd905dc677a6078990aad3331a36,5bf226cf882c5bf4247f89a97c86f273,1,NaN,O pedido consta de 2 produtos e até agora rece...,2017-09-09 00:00:00,2017-09-13 09:52:44
96080,0254bd905dc677a6078990aad3331a36,331b367bdd766f3d1cf518777317b5d9,1,NaN,O pedido consta de 2 produtos e até agora rece...,2017-09-09 00:00:00,2017-09-13 09:52:44


review_id
2d6ac45f859465b5c185274a1c929637    3
70509c441d994fa03d6c1457930c9024    3
9e25d6e3025e9b9a0fc7f03588d33e2b    3
08528f70f579f0c830189efc523d2182    3
f4bb9d6dd4fb6dcc2298f0e7b17b8e1e    3
ddc52555ca27b0fe67d5255147682d2d    3
e44840754f12fad2b8646712121b349a    3
4548534449b1f572e357211b90724f1b    3
4219a80ab469e3fc9901437b73da3f75    3
69a1068c3128a14994e3e422e4539e04    3
Name: order_id, dtype: int64


## Geolocation

In [6]:
geo = pd.read_csv(RAW_DIR / "olist_geolocation_dataset.csv")

print("Total rows:", len(geo))
print("Exact duplicate rows:", geo.duplicated().sum())
print(
    "Unique ZIP prefixes:",
    geo["geolocation_zip_code_prefix"].nunique()
)

Total rows: 1000163
Exact duplicate rows: 261831
Unique ZIP prefixes: 19015


# Findings

- All 9 Olist CSV files loaded successfully.
- Most expected key fields were unique.
- `review_id` is not unique in the raw reviews file:
  - 789 review IDs are duplicated.
  - Some are linked to multiple `order_id` values.
  - No cleaning decision has been made yet.
- The geolocation table has heavy duplication:
  - 1,000,163 rows
  - 261,831 exact duplicate rows
  - 19,015 unique ZIP prefixes
- Review comment fields contain substantial missing data.
- Some order delivery/approval dates and product fields are missing.
- Date/time fields are currently stored as strings and will be converted later.
- No raw data were modified during this inspection.


# Data Planning

### Core Tables
- Orders — order timing, delivery performance, customer link
- Order Items — seller, product, price, freight
- Reviews — customer satisfaction
- Sellers — seller location
- Products — category and product characteristics
- Customers — customer location

### Supporting Tables
- Category Translation — English product categories
- Geolocation — geographic coordinates; requires preparation before joining

### Optional
- Payments — may be used later if payment type/installments add value

### Planned Derived Variables
- delivery_days
- delivery_delay_days
- late_delivery_flag
- freight_percentage
- order_value
- seller_order_count
- seller_items_sold
- seller_total_sales
- seller_average_review
- seller_late_delivery_rate
- seller_category_count